server/src/controllers/grades.ts

```diff
@@ -88,3 +95,324 @@ export async function addGrades(req: Request, res: Response, next: NextFunction)
   // Close the readable stream once data reading finished.
   parser.end();
 }
+
+/**
+ * Recursively calculates the grade for a particular attainment by its formula
+ * node.
+ *
+ * TODO: Save calculated grades to the database.
+ */
+async function calculateFormulaNode(
+  formulaNode: FormulaNode,
+  presetGrades: Map<FormulaNode, number>
+): Promise<GradingResult> {
+  /*
+   * If a teacher has manually specified a grade for this attainment,
+   * the manually specified grade will be used.
+   *
+   * A grade has to be manually specified when using the 'Manual' formula
+   * and a grade may be manually specified for overriding grade calculation
+   * functions in special cases.
+   */
+  const presetGrade: number | undefined = presetGrades.get(formulaNode);
+  if (presetGrade) {
+    return {
+      status: Status.Pass,
```

Comment:
Can the teacher tell the system that a student has in fact failed an assignment?

Reply:
True, that should be possible in the future, but I think we can leave it out of this PR.